
package loamstream.compiler

/**
 * LoamStream
 * Created by oliverr on 10/6/2015.
 */
object PipelineTargetedPart07 {
  val string =
 """table path file project_epacts_ready_file=@project.epacts.ready dir project_epacts_dir disp ".epacts.ready"  class_level project comment "Dummy file to indicate ready to run epacts"

project_epacts_trunk=@project.epacts dir project_epacts_dir

table file path project_kinship_file=$project_epacts_trunk.kinf dir project_dir disp ".kinf" parent cat_project_sample_epacts_ibd_data class_level project comment "Kinship matrix for samples --- computed by EMMAX"

!!expand:,:project,type:project,all:project,sample:project_sample_subset,sample! \
project_type_marker_smart_pca_trunk=$project_sample_marker_plink_file.type.pca dir project_dir

file path project_sample_marker_smart_pca_map_file=$project_sample_marker_smart_pca_trunk.map dir project_dir disp ".pca.map" parent cat_project_sample_eigen_ibd_data class_level project comment "Map file for smartpca for description samples (maps project ID to pca ID)"

!!expand:,:filetype,Filetype:poplist,Poplist! \
!!expand:,:project,type,description:project,all,all:project,sample,project! \
file path project_type_marker_smart_pca_filetype_file=$project_type_marker_smart_pca_trunk.filetype dir project_dir disp ".pca.filetype" parent cat_project_type_eigen_ibd_data class_level project comment "Filetype file for smartpca for description samples"

!!expand:,:filetype,Filetype:evec,Evec:eval,Eval:weights,Weidht:log,Log:fam,Fam! \
!!expand:,:project,type,description:project,all,all:project,sample,project:project_sample_subset,sample,project_sample_subset! \
file path project_type_marker_smart_pca_filetype_file=$project_type_marker_smart_pca_trunk.filetype dir project_dir disp ".pca.filetype" parent cat_project_type_eigen_ibd_data class_level project comment "Filetype file for smartpca for description samples"

!!expand:project:project:project_sample_subset! \
table file path project_sample_marker_mds_file=$project_sample_marker_plink_file.mds dir project_dir disp ".mds" parent cat_project_sample_plink_ibd_data class_level project comment "MDS file for samples in project"

file path project_sample_marker_mds_log_file=$project_sample_marker_plink_file.mds.log dir project_dir disp ".mds.log" parent cat_project_sample_plink_ibd_data class_level project comment "Log file for mds cmd"

file path project_marker_custom_region_exclude_file=@project.sample.marker.custom.reg.exclude dir project_dir disp ".custom.reg.exclude" parent cat_project_region_exclude_ibd_data class_level project comment "Additional regions to exclude when computing genome/MDS"

file path project_marker_region_exclude_file=@project.sample.marker.all.reg.exclude dir project_dir disp ".reg.exclude" parent cat_project_region_exclude_ibd_data class_level project comment "Additional regions to exclude when computing genome/MDS"

#coverage metrics

!!expand:project:project:pheno! \
file path project_sample_subset_bam_list_file=@project_sample_subset.bam.list dir project_sample_subset_dir disp ".bam.list" parent cat_project_sample_subset_coverage_data class_level project_sample_subset comment "Bam list to feed into coverage computation for this collection of samples"

!!expand:project:project:pheno! \
table file path project_sample_subset_coverage_file=@project_sample_subset.coverage.txt.gz dir project_sample_subset_dir disp ".coverage.txt.gz" parent cat_project_sample_subset_coverage_data class_level project_sample_subset comment "Coverage at each base for this collection of samples; generated by DepthOfCoverageWalker"

!!expand:type:gene:exon:bait! \
file project_sample_type_coverage_stats_file_list=@project.sample.type.coverage.stats.list disp ".sample.stats.list" dir project_dir parent cat_project_type_dist_coverage_data class_level project comment "List of sample type coverage stats file for use in generating project type dist coverage dat file" child_order -1

file project_sample_coverage_file_list=@project.sample.coverage.list disp ".sample.coverage.list" dir project_dir parent cat_project_sample_coverage_data class_level project comment "List of sample coverage files" child_order -1

!!expand:type:gene:exon:bait! \
table file path project_type_dist_coverage_dat_file=@project.type.coverage.csv dir project_dir disp ".csv" parent cat_project_alphabetical_type_dist_coverage_data class_level project comment "Coverage for each sample stratified by type"
major file path project_gene_dist_coverage_pdf_file=@project.gene.coverage.pdf dir project_dir disp ".pdf" parent cat_project_alphabetical_gene_dist_coverage_data class_level project comment "Plot of coverage for each sample stratified by gene -- sorted alphabetically by gene name"

table file path project_lowest_gene_dist_coverage_dat_file=@project.gene.lowest.coverage.csv dir project_dir disp ".csv" parent cat_project_lowest_gene_dist_coverage_data class_level project comment "Coverage for each sample stratified by gene -- with a column for mean gene value"
major file path project_lowest_gene_dist_coverage_pdf_file=@project.gene.lowest.coverage.pdf dir project_dir disp ".pdf" parent cat_project_lowest_gene_dist_coverage_data class_level project comment "Plot of coverage for each sample stratified by gene -- sorted by lowest mean coverage"

!!expand:type:gene:exon:bait! \
table file path project_type_dist_mean_coverage_dat_file=@project.type.coverage.mean.csv dir project_dir disp ".mean.csv" parent cat_project_type_cum_coverage_data class_level project comment "Mean coverage for each sample stratified by type"

!!expand:type:gene:exon:bait! \
table nohead file path project_type_cum_coverage_dat_file=@project.type.cum.coverage.csv dir project_dir disp ".csv" parent cat_project_type_cum_coverage_data class_level project comment "Cumulative coverage distribution for all types"

!!expand:type:gene:exon:bait! \
major file path project_type_cum_coverage_pdf_file=@project.type.cum.coverage.pdf dir project_dir disp ".pdf" parent cat_project_type_cum_coverage_data class_level project comment "Plot of cumulative coverage distribution for all types"

file project_sample_coverage_stats_file_list=@project.sample.coverage.stats.list disp ".sample.list" dir project_dir parent cat_project_ind_sample_coverage_data class_level project comment "List of sample coverage stats file for use in generating project sample coverage dat file"

table file path project_sample_coverage_dat_file=@project.coverage.csv dir project_dir disp ".coverage.csv" parent cat_project_ind_sample_coverage_data class_level project comment "Coverage values for each passed sample"

table nohead file path project_sample_cum_coverage_dat_file=@project.sample.cum.coverage.csv dir project_dir disp ".csv" parent cat_project_sample_cum_coverage_data class_level project comment "Cumulative coverage for all samples --- includes failed"
major file path project_sample_cum_coverage_pdf_file=@project.sample.cum.coverage.pdf dir project_dir disp ".pdf" parent cat_project_sample_cum_coverage_data class_level project comment "Cumulative coverage for all samples --- includes failed"

#seq_batch

table file path seq_batch_picard_sort_file=@seq_batch.picard.sort.tsv dir seq_batch_dir disp ".picard.sort.tsv" parent cat_seq_batch_picard_data class_level seq_batch comment "Helper file mapping possible seq_batchtype values in the picard file to a sort order. The values in the picard file may not be the internal or disp seq_batch value; the entries in this file pertain to those in the picard file"

table file path seq_batch_picard_map_file=@seq_batch.picard.map dir seq_batch_dir disp ".picard.map" parent cat_seq_batch_picard_data class_level seq_batch comment "Helper file mapping unique ids to their batch"

table file path seq_batch_picard_file=@seq_batch.picard.tsv dir seq_batch_dir disp ".picard.tsv" parent cat_seq_batch_picard_data class_level seq_batch comment "Augmented picard file with batch information"


file path seq_batch_picard_pdf_file=@seq_batch.picard.metrics.pdf dir seq_batch_dir disp ".picard.metrics.pdf" parent cat_seq_batch_picard_data class_level seq_batch comment "Plot of select picard metrics stratified by seq_batch value in the picard file"

#pheno

!!expand:ktype:keep:exclude! \
table onecol nohead path file pheno_sample_ktype_file=@pheno.sample.ktype dir pheno_dir disp ".sample.ktype" parent cat_pheno_passed_meta_data class_level pheno comment "List of samples to ktype if meta_strata and meta_strata_value are specified"

table nohead path file pheno_sample_pheno_file=@pheno.sample.pheno dir pheno_dir disp ".sample.pheno" parent cat_pheno_passed_meta_data class_level pheno comment "Custom way to define phenotypes for samples"

table nohead path file pheno_non_missing_sample_pheno_file=@pheno.non_missing.sample.pheno dir pheno_dir disp ".non_missing.sample.pheno" parent cat_pheno_passed_meta_data class_level pheno comment "List of samples with non-missing phenotypes and their values for this phenotype --- only those that passed"

table nohead path file pheno_non_missing_sample_info_file=@pheno.non_missing.sample.info dir pheno_dir disp ".non_missing.sample.info" parent cat_pheno_passed_meta_data class_level pheno comment "List of samples with non-missing phenotypes and their values + disp values for this phenotype --- only those that passed"
table nohead path file pheno_missing_sample_info_file=@pheno.missing.sample.info dir pheno_dir disp ".missing.sample.info" parent cat_pheno_passed_meta_data class_level pheno comment "List of samples with missing phenotypes and their values for this phenotype --- passed samples"
doubcom table path file pheno_all_sample_info_file=@pheno.all.sample.info dir pheno_dir disp ".all.sample.info" parent cat_pheno_passed_meta_data class_level pheno comment "List of all samples and their values for this phenotype --- passed samples"

doubcom table path file pheno_plinkseq_indicator_phe_file_body=@pheno.plinkseq.indicator.phe.body dir pheno_dir disp ".indicator.phe.body" parent cat_pheno_passed_meta_data class_level pheno comment "Phe body file with indicator variables"

table nohead path file pheno_failed_non_missing_sample_pheno_file=@pheno.failed.non_missing.sample.pheno dir pheno_dir disp ".non_missing.sample.pheno" parent cat_pheno_failed_meta_data class_level pheno comment "List of samples with non-missing phenotypes and their values for this phenotype --- only those that failed"
table nohead path file pheno_failed_non_missing_sample_info_file=@pheno.failed.non_missing.sample.info dir pheno_dir disp ".non_missing.sample.info" parent cat_pheno_failed_meta_data class_level pheno comment "List of samples with non-missing phenotypes and their values + disp values for this phenotype --- only those that failed"
table nohead path file pheno_failed_missing_sample_info_file=@pheno.failed.missing.sample.info dir pheno_dir disp ".missing.sample.info" parent cat_pheno_failed_meta_data class_level pheno comment "List of samples with missing phenotypes and their values for this phenotype --- failed samples"

minor path file pheno_sample_info_header_file=@pheno.sample.info.header dir pheno_dir disp ".sample.info.header" parent cat_pheno_all_meta_data class_level pheno comment "Header for sample info file"

minor path file pheno_indicator_list_file=@pheno.indicator.list dir pheno_dir disp ".indicator.list" parent cat_pheno_all_meta_data class_level pheno comment "List of indicator values"

minor path file pheno_variable_covars_file=@pheno.variable.covars dir pheno_dir disp ".variable.covars" parent cat_pheno_all_meta_data class_level pheno comment "List of covariates that are variable in these samples"

minor path file pheno_plinkseq_indicator_phe_file_header1=@pheno.plinkseq.indicator.phe.header1 dir pheno_dir disp ".indicator.phe.header1" parent cat_pheno_all_meta_data class_level pheno comment "Header for sample indicator file"

minor path file pheno_disp_order_file=@pheno.disp.order.tsv dir pheno_dir disp ".disp.order" parent cat_pheno_all_meta_data class_level pheno comment "Simple map from disp to order\n"

table path file pheno_quantile_file=@pheno.quantile dir pheno_dir disp ".quantile" parent cat_pheno_all_meta_data class_level pheno comment "List of quantile for this phenotype; grouped for passed, failed, and all samples"

doubcom table path file pheno_sequenced_all_sample_info_file=@pheno.sequenced.all.sample.info dir pheno_dir disp ".all.sample.info" parent cat_pheno_all_meta_data class_level pheno comment "List of all samples and their values for this phenotype --- passed and failed samples"

doubcom table path file pheno_sequenced_all_sample_box_info_file=@pheno.all.sample.box.info dir pheno_dir disp ".all.sample.box.info" parent cat_pheno_all_meta_data class_level pheno comment "Trait values for each sample, ready for box plotting"

doubcom table path file pheno_sequenced_all_trait_all_sample_box_info_file=@pheno.all_trait.all.sample.info dir pheno_dir disp ".all_trait.all.sample.info" parent cat_pheno_all_meta_data class_level pheno comment "List of all samples and their values for this as well as related phenotypes --- passed and failed samples"

table path file pheno_sample_failure_status_file=@pheno.samples.failure.status dir pheno_dir disp ".failure.status" parent cat_pheno_all_meta_data class_level pheno comment "Columns for each fail/pass designation; like project level file but with additional ibd removal information for this phenotype"

doubcom table file path pheno_cross_classification_phe_file=@pheno.cross.phe dir pheno_dir disp ".cross.phe" parent cat_pheno_all_meta_data class_level pheno comment "Modified phenotype file with two differences: (1) Only have traits suitable for cross classification as specified in meta file and (2) Showing display names rather than internal names"


minor file path pheno_plinkseq_phe_file_header1=@pheno.plinkseq.phe.header1 dir pheno_dir disp ".phe.header1" parent cat_pheno_plinkseq_phe_data class_level pheno comment "First header lines for Plink/Seq phe file"

minor file path pheno_plinkseq_phe_file_body=@pheno.plinkseq.phe.body dir pheno_dir disp ".phe.body" parent cat_pheno_plinkseq_phe_data class_level pheno comment "Plink/Seq phe body file"

file table doubcom path pheno_plinkseq_phe_file=@pheno.plinkseq.phe dir pheno_dir disp ".phe" parent cat_pheno_plinkseq_phe_data class_level pheno comment "Plink/Seq phe file"

minor file path pheno_plinkseq_strata_phe_file_header1=@pheno.plinkseq.strata.phe.header1 dir pheno_dir disp ".strata.phe.header1" parent cat_pheno_plinkseq_phe_data class_level pheno comment "First header lines for Plink/Seq strata phe file"
minor file path pheno_plinkseq_strata_phe_file_body=@pheno.plinkseq.strata.phe.body dir pheno_dir disp ".strata.phe.body" parent cat_pheno_plinkseq_phe_data class_level pheno comment "Body lines for Plink/Seq strata phe file"
doubcom table file path pheno_plinkseq_strata_phe_file=@pheno.plinkseq.strata.phe dir pheno_dir disp ".strata.phe" parent cat_pheno_plinkseq_phe_data class_level pheno comment "Strata phenotype file for import into Plink/Seq"

doubcom table file path pheno_plinkseq_covar_phe_file=@pheno.covar.phe dir pheno_dir disp ".covar.phe" parent cat_pheno_plinkseq_phe_data class_level pheno comment "Additional covar phenotype file for import into Plink/Seq"

doubcom table file path pheno_plinkseq_cluster_phe_file=@pheno.cluster.phe dir pheno_dir disp ".cluster.phe" parent cat_pheno_plinkseq_phe_data class_level pheno comment "Additional cluster phenotype file for import into Plink/Seq"

#table file path pheno_score_seq_phe_file=@pheno.score_seq.phe dir pheno_dir disp ".score_seq.phe" parent cat_pheno_plinkseq_phe_data class_level pheno comment "Phenotypes in format for Score/SEQ"

#table file path pheno_score_seq_phe_covar_file=@pheno.score_seq.phe.covar dir pheno_dir disp ".score_seq.phe.covar" parent cat_pheno_plinkseq_phe_data class_level pheno comment "Phenotypes + covariates in format for Score/SEQ"

#table file path pheno_score_seq_reverse_phe_file=@pheno.score_seq.rev.phe dir pheno_dir disp ".score_seq.rev.phe" parent cat_pheno_plinkseq_phe_data class_level pheno comment "Phenotypes in format for Score/SEQ --- case/ctrl flipped"

#table file path pheno_score_seq_reverse_phe_covar_file=@pheno.score_seq.rev.phe.covar dir pheno_dir disp ".score_seq.rev.phe.covar" parent cat_pheno_plinkseq_phe_data class_level pheno comment "Phenotypes + covariates in format for Score/SEQ --- case/ctrl flipped"

!!expand:type:all:unrelated! \
onecol table nohead path file pheno_sample_non_missing_type_include_file=@pheno.sample.non_missing.type.include dir pheno_dir disp ".sample.non_missing.type.include" parent cat_pheno_plinkseq_type_include_data class_level pheno comment "Samples (type) included for analysis in this phenotype (passing QC + non missing phenotype)"

!!expand:type:all:unrelated! \
onecol table nohead path file pheno_sample_basic_type_include_file=@pheno.sample.basic.type.include dir pheno_dir disp ".sample.basic.type.include" parent cat_pheno_plinkseq_type_include_data class_level pheno comment "Samples (type) included for analysis in this phenotype (passing QC + non missing phenotype)"

!!expand:type:all:unrelated! \
onecol table nohead path file pheno_sample_popgen_type_include_file=@pheno.sample.popgen.type.include dir pheno_dir disp ".sample.popgen.type.include" parent cat_pheno_plinkseq_type_include_data class_level pheno comment "Samples (type) included for analysis in this phenotype (passing QC + Popgen)"

#jason onecol table file path pheno_sample_filtered_mds_include_file=@pheno.sample.filtered.mds.include dir pheno_dir disp ".sample.filtered.mds.include" parent cat_pheno_plinkseq_include_data class_level pheno comment "Samples included due to outliers on MDS scale"

!!expand:type:all:unrelated! \
!!expand:,:filetype,description:covar,not removed during MDS computation or during MDS outlier evaluation:cluster,not removed during MDS computation or MDS outlier evaluation or clustering! \
onecol table nohead file path pheno_sample_filetype_type_include_file=@pheno.sample.filetype.type.include dir pheno_dir disp ".sample.filetype.type.include" parent cat_pheno_plinkseq_type_include_data class_level pheno comment "Samples (type) to include during filetype evaluation: description"


!!expand:pheno:pheno:pheno_variant_subset! \
constant pheno_project_exon_browser=Open disp "Exon Browser" parent cat_pheno_plinkseq_raw_data goto_url ${plinkseq_exome_browser_url}?proj=*pheno_plinkseq_project_file&loc=$pseq_genes_loc_group comment "Link to Plink/Seq exon browser for raw pheno project"
!!expand:pheno:pheno:pheno_variant_subset! \
table nohead file path pheno_plinkseq_project_file=@pheno.plinkseq.project.pseq dir pheno_dir disp ".plinkseq.project" parent cat_pheno_plinkseq_raw_data class_level pheno comment "Plink/Seq raw pheno project file"
!!expand:pheno:pheno:pheno_variant_subset! \
mkdir path pheno_plinkseq_project_out_dir=*{pheno_plinkseq_project_file}_out class_level pheno chmod 777
!!expand:pheno:pheno:pheno_variant_subset! \
mkdir path pheno_plinkseq_project_temp_dir=*{pheno_plinkseq_project_file}_temp class_level pheno chmod 777
!!expand:pheno:pheno:pheno_variant_subset! \
file path pheno_plinkseq_inddb_file=inddb dir pheno_plinkseq_project_out_dir disp "inddb" parent cat_pheno_plinkseq_raw_data class_level pheno comment "Plink/Seq raw pheno inddb file"
!!expand:pheno:pheno:pheno_variant_subset! \
file path pheno_plinkseq_project_done_file=.$pheno_plinkseq_project_file.done dir pheno_dir disp "project.done" parent cat_pheno_plinkseq_raw_data class_level pheno comment "Signifies that raw pheno project was done"
!!expand:pheno:pheno:pheno_variant_subset! \
ignore_md5 file path pheno_plinkseq_db_done_file=db.done dir pheno_plinkseq_project_out_dir disp "db.done" parent cat_pheno_plinkseq_raw_data class_level pheno comment "Signifies that raw pheno db was done"
!!expand:pheno:pheno:pheno_variant_subset! \
file path pheno_plinkseq_temp_done_file=temp.done dir pheno_plinkseq_project_temp_dir disp "temp.done" parent cat_pheno_plinkseq_raw_data class_level pheno comment "Signifies that raw pheno temp dir was done"

!!expand:pheno:pheno:pheno_variant_subset! \
constant pheno_clean_project_exon_browser=Open disp "Exon Browser" parent cat_pheno_plinkseq_clean_data goto_url ${plinkseq_exome_browser_url}?proj=*pheno_plinkseq_clean_project_file&loc=$pseq_genes_loc_group comment "Link to Plink/Seq exon browser for clean pheno project"
!!expand:pheno:pheno:pheno_variant_subset! \
table nohead file path pheno_plinkseq_clean_project_file=@pheno.plinkseq.clean.project.pseq dir pheno_dir disp ".plinkseq.project" parent cat_pheno_plinkseq_clean_data class_level pheno comment "Plink/Seq clean pheno project file"
!!expand:pheno:pheno:pheno_variant_subset! \
mkdir path pheno_plinkseq_clean_project_out_dir=*{pheno_plinkseq_clean_project_file}_out class_level pheno chmod 777
!!expand:pheno:pheno:pheno_variant_subset! \
mkdir path pheno_plinkseq_clean_project_temp_dir=*{pheno_plinkseq_clean_project_file}_temp class_level pheno chmod 777
!!expand:pheno:pheno:pheno_variant_subset! \
file path pheno_plinkseq_clean_inddb_file=inddb dir pheno_plinkseq_clean_project_out_dir disp "inddb" parent cat_pheno_plinkseq_clean_data class_level pheno comment "Plink/Seq clean pheno inddb file"
!!expand:pheno:pheno:pheno_variant_subset! \
file path pheno_plinkseq_clean_project_done_file=.$pheno_plinkseq_clean_project_file.done dir pheno_dir disp "project.done" parent cat_pheno_plinkseq_clean_data class_level pheno comment "Signifies that clean pheno project was done"
!!expand:pheno:pheno:pheno_variant_subset! \
ignore_md5 file path pheno_plinkseq_clean_db_done_file=db.done dir pheno_plinkseq_clean_project_out_dir disp "db.done" parent cat_pheno_plinkseq_clean_data class_level pheno comment "Signifies that clean pheno db was done"
!!expand:pheno:pheno:pheno_variant_subset! \
file path pheno_plinkseq_clean_temp_done_file=temp.done dir pheno_plinkseq_clean_project_temp_dir disp "temp.done" parent cat_pheno_plinkseq_clean_data class_level pheno comment "Signifies that clean pheno temp dir was done"

!!expand:pheno:pheno:pheno_variant_subset! \
constant pheno_strat_project_exon_browser=Open disp "Exon Browser" parent cat_pheno_plinkseq_strat_data goto_url ${plinkseq_exome_browser_url}?proj=*pheno_plinkseq_strat_project_file&loc=$pseq_genes_loc_group comment "Link to Plink/Seq exon browser for strat pheno project"
!!expand:pheno:pheno:pheno_variant_subset! \
table nohead file path pheno_plinkseq_strat_project_file=@pheno.plinkseq.strat.project.pseq dir pheno_dir disp ".plinkseq.project" parent cat_pheno_plinkseq_strat_data class_level pheno comment "Plink/Seq strat pheno project file"
!!expand:pheno:pheno:pheno_variant_subset! \
mkdir path pheno_plinkseq_strat_project_out_dir=*{pheno_plinkseq_strat_project_file}_out class_level pheno chmod 777
!!expand:pheno:pheno:pheno_variant_subset! \
mkdir path pheno_plinkseq_strat_project_temp_dir=*{pheno_plinkseq_strat_project_file}_temp class_level pheno chmod 777
!!expand:pheno:pheno:pheno_variant_subset! \
file path pheno_plinkseq_strat_inddb_file=inddb dir pheno_plinkseq_strat_project_out_dir disp "inddb" parent cat_pheno_plinkseq_strat_data class_level pheno comment "Plink/Seq strat pheno inddb file"
!!expand:pheno:pheno:pheno_variant_subset! \
file path pheno_plinkseq_strat_project_done_file=.$pheno_plinkseq_strat_project_file.done dir pheno_dir disp "project.done" parent cat_pheno_plinkseq_strat_data class_level pheno comment "Signifies that strat pheno project was done"
!!expand:pheno:pheno:pheno_variant_subset! \
ignore_md5 file path pheno_plinkseq_strat_db_done_file=db.done dir pheno_plinkseq_strat_project_out_dir disp "db.done" parent cat_pheno_plinkseq_strat_data class_level pheno comment "Signifies that strat pheno db was done"
!!expand:pheno:pheno:pheno_variant_subset! \
file path pheno_plinkseq_strat_temp_done_file=temp.done dir pheno_plinkseq_strat_project_temp_dir disp "temp.done" parent cat_pheno_plinkseq_strat_data class_level pheno comment "Signifies that strat pheno temp dir was done"

table nohead file path pheno_plink_phe_file=@pheno.plink.phe dir pheno_dir disp ".plink.phe" parent cat_pheno_plink_phe_data class_level pheno comment "Phenotype file for loading into plink"

table file path pheno_plink_alternate_phe_file=@pheno.plink.alternate.phe dir pheno_dir disp ".plink.alternate.phe" parent cat_pheno_plink_phe_data class_level pheno comment "Phenotype file for loading into plink --- has all traits in plinkseq pheno file"

table file path pheno_plink_covar_file=@pheno.plink.covar dir pheno_dir disp ".plink.covar" parent cat_pheno_plink_phe_data class_level pheno comment "Covar file for loading into plink"

table file path pheno_plink_covar_cluster_file=@pheno.plink.covar.cluster dir pheno_dir disp ".plink.covar.cluster" parent cat_pheno_plink_phe_data class_level pheno comment "Covar file for loading into plink (has cluster info as well)"

!!expand:type:all:unrelated! \
table file path pheno_epacts_type_ped_file=@pheno.epacts.type.ped dir pheno_dir disp ".epacts.type.ped" parent cat_pheno_epacts_phe_data class_level pheno comment "Phenotype file for loading into EPACTS --- includes all samples samples with non-missing phenotypes"

!!expand:type:all:unrelated! \
table file path pheno_epacts_popgen_type_ped_file=@pheno.epacts.popgen.type.ped dir pheno_dir disp ".epacts.popgen.type.ped" parent cat_pheno_epacts_phe_data class_level pheno comment "Phenotype file for loading into EPACTS --- includes popgen+ samples only"

!!expand:type:all:unrelated! \
table file path pheno_epacts_covar_type_ped_file=@pheno.epacts.covar.type.ped dir pheno_dir disp ".epacts.covar.type.ped" parent cat_pheno_epacts_phe_data class_level pheno comment "Phenotype file for loading into EPACTS --- includes covar+ samples only"

!!expand:type:all:unrelated! \
onecol table nohead path file pheno_sample_plink_non_missing_type_include_file=@pheno.sample.plink.non_missing.type.include dir pheno_dir disp ".sample.plink.non_missing.type.include" parent cat_pheno_plink_type_include_data class_level pheno comment "Samples (type) included for analysis in this phenotype (passing QC+ with non-missing phenotype)"

!!expand:type:all:unrelated! \
onecol table nohead path file pheno_sample_plink_basic_type_include_file=@pheno.sample.plink.basic.type.include dir pheno_dir disp ".sample.plink.basic.type.include" parent cat_pheno_plink_type_include_data class_level pheno comment "Samples (type) included for analysis in this phenotype (passing QC+ with non-missing phenotype)"

!!expand:type:all:unrelated! \
onecol table nohead path file pheno_sample_plink_popgen_type_include_file=@pheno.sample.plink.popgen.type.include dir pheno_dir disp ".sample.plink.popgen.type.include" parent cat_pheno_plink_type_include_data class_level pheno comment "Samples (type) included for analysis in this phenotype (passing QC + Popgen)"

#onecol table nohead path file pheno_sample_plink_filtered_mds_type_include_file=@pheno.sample.plink.filtered.mds.include dir pheno_dir disp ".sample.plink.filtered.mds.include" parent cat_pheno_plink_include_data class_level pheno comment "Samples included due to outliers on MDS scale"

!!expand:type:all:unrelated! \
!!expand:filetype:covar:cluster! \
onecol table nohead file path pheno_sample_plink_filetype_type_include_file=@pheno.sample.plink.filetype.type.include dir pheno_dir disp ".sample.plink.filetype.type.include" parent cat_pheno_plink_type_include_data class_level pheno comment "Samples (type) to include during filetype evaluation"


table file path pheno_sstats_file=@pheno.sstats dir pheno_dir disp ".sstats" parent cat_pheno_sstats_qc_stats_data class_level pheno comment "Project level sstats projected to this phenotype"

major file path pheno_trait_hist_pdf_file=@pheno.trait.hist.pdf dir pheno_dir disp ".trait.hist.pdf" parent cat_pheno_slide_sample_qc_data class_level pheno comment "Histogram of sample trait values"

!!expand:,:slidedot,slideunder,cattouse,extracomment:slide.,slide_,cat_pheno_slide_sample_qc_data,-- fewer columns for inclusion in slides:,,cat_pheno_phenocat_data,! \
!!expand:,:tval,tname,phenocat\
:all,all,sstats_qc_plot\
:highlighted,highlighted outlier,sstats_qc_plot\
:filtered,QC+,sstats_qc_plot\
:extreme,QC+,sstats_qc_plot\
:popgen,QC/Popgen+,popgen_qc_plot!\
major file path pheno_slideundertval_sstats_pdf_file=@pheno.slidedottval.sstats.pdf dir pheno_dir disp ".tval.sstats.pdf" parent cattouse class_level pheno comment "Plot of various sample metrics across all samples stratified by phenotype value; tname samples extracomment"

#pheno_sample_marker_type_nearest_file defined above

table file path pheno_project_mds_file=@pheno.project.mds dir pheno_dir disp ".project.mds" parent cat_pheno_covar_qc_stats_data class_level pheno comment "Project level MDS values projected to this phenotype --- only includes QC+ samples"

table file path pheno_mds_exclude_file=@pheno.mds.exclude dir pheno_dir disp ".mds.exclude" parent cat_pheno_plinkseq_all_include_data class_level pheno comment "Samples excluded due to outliers according to mds"

onecol table nohead path file pheno_basic_exclude_file=@pheno.sample.basic.exclude dir pheno_dir disp ".sample.basic.exclude" parent cat_pheno_plinkseq_all_include_data class_level pheno comment "Exclude these samples from any phenotype specific analysis: potential combination of MDS outliers, sex check discordances, duplicates"

!!expand:pheno:pheno:pheno_sample_subset! \
!!expand:,:type,parent_cat\
:all,cat_pheno_all_qc_stats_data\
:unrelated,cat_pheno_unrelated_qc_stats_data! \
pheno_sample_marker_type_nearest_trunk=@pheno.type dir pheno_dir

!!expand:,:type,parent_cat\
:all,cat_pheno_popgen_qc_stats_data\
:unrelated,cat_pheno_popgen_qc_stats_data! \
table file path pheno_sample_marker_type_nearest_file=$pheno_sample_marker_type_nearest_trunk.nearest dir pheno_dir disp ".type.nearest" parent parent_cat class_level pheno comment "Nearest file for type QC+ samples in project"

!!expand:,:type,parent_cat\
:all,cat_pheno_popgen_qc_stats_data\
:unrelated,cat_pheno_popgen_qc_stats_data! \
file path pheno_sample_marker_type_nearest_log_file=@pheno.type.nearest.log dir pheno_dir disp ".type.nearest.log" parent parent_cat class_level pheno comment "Log file for type nearest cmd"

!!expand:pheno:pheno:pheno_sample_subset! \
!!expand:,:type,removal_t,cat_t:all,no,popgen:unrelated,popgen- and related,popgen! \
table file path pheno_type_popgen_istats_file=@pheno.type.popgen.istats dir pheno_dir disp ".type.popgen.istats" parent cat_pheno_cat_t_qc_stats_data class_level pheno comment "Sample statistics based on popgen metrics --- bad samples removed; removal_t samples removed"

major table file path pheno_popgen_outlier_file=@pheno.popgen.outliers.csv dir pheno_dir disp ".popgen.outliers.csv" parent cat_pheno_popgen_qc_stats_data class_level pheno comment "Outlier values from popgen statistics"

onecol table file path pheno_additional_popgen_exclude_file=@pheno.additional.popgen.exclude dir pheno_dir disp ".additional.popgen.exclude" parent cat_pheno_popgen_qc_stats_data class_level pheno comment "Custom samples to exclude; will exclude from MDS computation as well"

table file path pheno_additional_popgen_plink_exclude_file=@pheno.additional.popgen.plink.exclude dir pheno_dir disp ".additional.popgen.plink.exclude" parent cat_pheno_popgen_qc_stats_data class_level pheno comment "Custom samples to exclude (plink format)"

table file path pheno_popgen_exclude_detail_file=@pheno.popgen.exclude.detail dir pheno_dir disp ".popgen.exclude.detail" parent cat_pheno_popgen_qc_stats_data class_level pheno comment "Samples excluded because they are outliers on popgen measures"

onecol table nohead path file pheno_popgen_exclude_file=@pheno.popgen.exclude dir pheno_dir disp ".popgen.exclude" parent cat_pheno_popgen_qc_stats_data class_level pheno comment "Samples excluded based on popgen measures"

table nohead path file pheno_related_exclude_custom_rank_file=@pheno.related.exclude.custom.rank dir pheno_dir disp ".related.exclude.custom.rank" parent cat_pheno_ibd_qc_stats_data class_level pheno comment "Custom rank samples to exclude: higher number means preferentially include"

table nohead path file pheno_related_exclude_rank_file=@pheno.related.exclude.rank dir pheno_dir disp ".related.exclude.rank" parent cat_pheno_ibd_qc_stats_data class_level pheno comment "Rank samples to exclude: higher number means preferentially include"


onecol table nohead path file pheno_custom_related_exclude_file=@pheno.custom.related.exclude dir pheno_dir disp ".custom.related.exclude" parent cat_pheno_ibd_qc_stats_data class_level pheno comment "Custom samples to exclude for relatedness"

onecol table nohead path file pheno_basic_related_exclude_file=@pheno.basic.related.exclude dir pheno_dir disp ".basic.related.exclude" parent cat_pheno_ibd_qc_stats_data class_level pheno comment "Samples excluded from basic include file based on relatedness"

!!expand:,:_type,.type,Type:_all,.all,all:_unrelated,.unrelated,unrelated! \
"""
}
    